### ISSUE TITLE
Play 2.7.0-RC9 fails to evolve database

### ISSUE THREAD

**Issue by Squeng** [playframework/playframework#8947]
might be related to https://github.com/playframework/playframework/issues/8935

### Play Version (2.5.x / etc)

2.7.0-RC9

### API (Scala / Java / Neither / Both)

Java

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)

Windows 10

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

java version "11.0.1" 2018-10-16 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.1+13-LTS)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.1+13-LTS, mixed mode)

### Expected Behavior

As documented at https://www.playframework.com/documentation/2.7.x/Evolutions#Evolutions-scripts

### Actual Behavior

```
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[NoSuchElementException: key not found: Id]]
        at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:323)
        at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:243)
        at play.api.http.DefaultHttpErrorHandler$.onServerError(HttpErrorHandler.scala:464)
        at play.core.server.Server$$anonfun$handleErrors$1$1.applyOrElse(Server.scala:96)
        at play.core.server.Server$$anonfun$handleErrors$1$1.applyOrElse(Server.scala:92)
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
        at play.core.server.Server$.getHandlerFor(Server.scala:115)
        at play.core.server.AkkaHttpServer.handleRequest(AkkaHttpServer.scala:267)
        at play.core.server.AkkaHttpServer.$anonfun$createServerBinding$1(AkkaHttpServer.scala:149)
        at akka.stream.impl.fusing.MapAsync$$anon$30.onPush(Ops.scala:1207)
Caused by: java.util.NoSuchElementException: key not found: Id
        at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:101)
        at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:99)
        at play.api.libs.typedmap.DefaultTypedMap.apply(TypedMap.scala:102)
        at play.api.mvc.RequestHeader.id(RequestHeader.scala:35)
        at play.api.mvc.RequestHeader.id$(RequestHeader.scala:35)
        at play.api.mvc.RequestHeaderImpl.id(RequestHeader.scala:352)
        at play.core.j.JavaHelpers.createJavaContext(JavaHelpers.scala:172)
        at play.core.j.JavaHelpers.createJavaContext$(JavaHelpers.scala:169)
        at play.core.j.JavaHelpers$.createJavaContext(JavaHelpers.scala:282)
        at play.core.j.JavaHelpers.withContext(JavaHelpers.scala:270)
```
### Reproducible Test Case

start with 1.sql and restart with 9.sql (after renaming it to 2.sql)

[sql.zip](https://github.com/playframework/playframework/files/2759721/sql.zip)

**Comment by mkurz** [comment]
Can you please provide more information? Are you running your application in dev mode?
Do you have a module which runs a query against the database?

**Comment by Squeng** [comment]
Yes, I'm running in DEV mode:
`[FiLe-App] $ run -Dhttp.port=9000 -Dhttps.port=9443`
with
`play.evolutions.autoApply=true`
in application.conf.

I've restarted my app with
`play.evolutions.autoApply=false`

So there is a workaround.

**Comment by mkurz** [comment]
Do you have any modules that access the database?

**Comment by Squeng** [comment]
Thanks for having edited my comment from before, @mkurz !

I'll try to remember to do so myself in the future.

**Comment by mkurz** [comment]
Hmm... after looking at the exception log maybe this is not releated to evolutions...
```
Caused by: java.util.NoSuchElementException: key not found: Id

at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:101)
at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:99)
at play.api.libs.typedmap.DefaultTypedMap.apply(TypedMap.scala:102)
at play.api.mvc.RequestHeader.id(RequestHeader.scala:35)
at play.api.mvc.RequestHeader.id$(RequestHeader.scala:35)
at play.api.mvc.RequestHeaderImpl.id(RequestHeader.scala:352)
at play.core.j.JavaHelpers.createJavaContext(JavaHelpers.scala:172)
at play.core.j.JavaHelpers.createJavaContext$(JavaHelpers.scala:169)
at play.core.j.JavaHelpers$.createJavaContext(JavaHelpers.scala:282)
at play.core.j.JavaHelpers.withContext(JavaHelpers.scala:270)
```

https://github.com/playframework/playframework/blob/cc8eacf3d6fe40fb8c91e692eef0729b852101a8/framework/src/play/src/main/scala/play/core/j/JavaHelpers.scala#L172

https://github.com/playframework/playframework/blob/cc8eacf3d6fe40fb8c91e692eef0729b852101a8/framework/src/play/src/main/scala/play/api/mvc/RequestHeader.scala#L35

Looks like there wasn't a request id...

@Squeng Please stop your application, remove the `target` folder and try again. Does it still happen?

**Comment by Squeng** [comment]
# first
I've re-set application.conf to say
`play.evolutions.autoApply=false`
(Note that I originally got it wrong in https://github.com/playframework/playframework/issues/8947#issuecomment-454391584)

I've added another SQL script:
```sql
# following up on https://github.com/playframework/playframework/issues/8947


# --- !Ups

ALTER TABLE Users ADD debugging bit;


# --- !Downs

ALTER TABLE Users DROP debugging;
```

Play 2.7.0-RC9 is back to failing:
```
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[NoSuchElementException: key not found: Id]]
        at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:323)
        at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:243)
        at play.api.http.DefaultHttpErrorHandler$.onServerError(HttpErrorHandler.scala:464)
        at play.core.server.Server$$anonfun$handleErrors$1$1.applyOrElse(Server.scala:96)
        at play.core.server.Server$$anonfun$handleErrors$1$1.applyOrElse(Server.scala:92)
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
        at play.core.server.Server$.getHandlerFor(Server.scala:115)
        at play.core.server.AkkaHttpServer.handleRequest(AkkaHttpServer.scala:267)
        at play.core.server.AkkaHttpServer.$anonfun$createServerBinding$1(AkkaHttpServer.scala:149)
        at akka.stream.impl.fusing.MapAsync$$anon$30.onPush(Ops.scala:1207)
Caused by: java.util.NoSuchElementException: key not found: Id
        at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:101)
        at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:99)
        at play.api.libs.typedmap.DefaultTypedMap.apply(TypedMap.scala:102)
        at play.api.mvc.RequestHeader.id(RequestHeader.scala:35)
        at play.api.mvc.RequestHeader.id$(RequestHeader.scala:35)
        at play.api.mvc.RequestHeaderImpl.id(RequestHeader.scala:352)
        at play.core.j.JavaHelpers.createJavaContext(JavaHelpers.scala:172)
        at play.core.j.JavaHelpers.createJavaContext$(JavaHelpers.scala:169)
        at play.core.j.JavaHelpers$.createJavaContext(JavaHelpers.scala:282)
        at play.core.j.JavaHelpers.withContext(JavaHelpers.scala:270)
```

# second
I've stopped the app (and exited the sbt console) and then removed the `target` folder as requested by https://github.com/playframework/playframework/issues/8947#issuecomment-454393874

# third
I've re-entered the sbt console (which resulted in a new `target` folder) and re-started the app.
It still fails with the exception above.

# fourth
I've deleted the new SQL script from both the `conf\evolutions\default` folder as well as within `target` and reloaded the page.
Everything is back to normal (i.e., the app is up & running).

**Comment by mkurz** [comment]
@Squeng would it be possible to provide us a test project so we can test ourselves? You could remove all stuff that is not relevant for this issue here, and upload to GitHub?

**Comment by Squeng** [comment]
@mkurz After having stripped the project of all that seemed irrelevant, Evolutions are working ...

I'll have to re-strip the project less aggressively; maybe I can figure out what the tipping point is.

**Comment by mkurz** [comment]
Interesting. Let us know what you find out along the way.

**Comment by Squeng** [comment]
@mkurz After quite some old-fashioned trial & error, I have realized that Evolutions work fine if I get rid of my `ErrorHandler`:
```java

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

import javax.inject.Inject;
import javax.inject.Provider;
import javax.inject.Singleton;

import com.typesafe.config.Config;

import play.Environment;
import play.Logger;
import play.api.OptionalSourceMapper;
import play.api.routing.Router;
import play.http.DefaultHttpErrorHandler;
import play.mvc.Http.RequestHeader;
import play.mvc.Result;
import play.mvc.Results;


@Singleton
public class ErrorHandler
    extends DefaultHttpErrorHandler
{
    // see also StrictTransportSecurityFilter

    private final static Logger.ALogger logger = Logger.of( "application" );

    private final String                httpPort;
    private final String                httpsPort;


    private String colonPort( String port )
    {
        if ( !port.isEmpty() ) {
            return ":" + port;
        } else {
            return "";
        }
    }

    @Inject
    public ErrorHandler( Config configuration,
                         Environment environment,
                         OptionalSourceMapper sourceMapper,
                         Provider<Router> routes )
    {
        super( configuration, environment, sourceMapper, routes );

        httpPort = colonPort( configuration.hasPath( "http.port" ) ? configuration.getString( "http.port" ) : "" );
        httpsPort = colonPort( configuration.hasPath( "https.port" ) ? configuration.getString( "https.port" ) : "" );
    }

    protected CompletionStage<Result> onNotFound( RequestHeader request, String message )
    {
        logger.debug( "path {} ({}) requested but not found", request.path(), request.secure() );

        if ( "/".equals( request.path() ) || "/filewaf".equals( request.path() ) || "/filewaf/".equals( request.path() ) ) {
            // assume that a user requested the root and refer her/him to the frontend
            String absoluteUrl = controllers.frontend.routes.FrontendController.index().absoluteURL( true,
                    request.host().replace( httpPort, httpsPort ) );
            logger.debug( "referring to absolute URL {} instead", absoluteUrl );
            return CompletableFuture.completedFuture( Results.temporaryRedirect( absoluteUrl ) );
        } else if ( "/filewaf/bankers".equals( request.path() ) || "/filewaf/frontend".equals( request.path() ) ) {
            String path = request.path() + "/";
            logger.debug( "referring to path {} instead", path );
            return CompletableFuture.completedFuture( Results.temporaryRedirect( path ) );
        } else if ( request.path().endsWith( "/" ) ) {
            String path = request.path().substring( 0, request.path().length() - 1 );
            logger.debug( "referring to path {} instead", path );
            return CompletableFuture.completedFuture( Results.temporaryRedirect( path ) );
        } else {
            return super.onNotFound( request, message );
        }
    }
}
```
Maybe it's because of the changes I made to it for 2.7.0-RC8:
![differrorhandler](https://user-images.githubusercontent.com/7399173/51263335-5b6a1f80-19b4-11e9-89f9-1b5ac81c0eae.png)

**Comment by mkurz** [comment]
Alright, can you debug which `if` branch is called and also again post the *whole* exception? Thanks!

**Comment by mkurz** [comment]
I mean the `onNotFound` in your error handler will even be called or not?

**Comment by Squeng** [comment]
@mkurz No, it's not. And the `ErrorHandler`is even causing trouble in the following stripped-down version and when I'm entering a URL (for robots.txt) that exists:
```java
import javax.inject.Inject;
import javax.inject.Provider;
import javax.inject.Singleton;

import com.typesafe.config.Config;

import play.Environment;
import play.api.OptionalSourceMapper;
import play.api.routing.Router;
import play.http.DefaultHttpErrorHandler;


@Singleton
public class ErrorHandler
    extends DefaultHttpErrorHandler
{
    @Inject
    public ErrorHandler( Config configuration,
                         Environment environment,
                         OptionalSourceMapper sourceMapper,
                         Provider<Router> routes )
    {
        super( configuration, environment, sourceMapper, routes );
    }
}
```
The log message and the stack trace are (the `Subsystem = play.api.http.DefaultHttpErrorHandler` part was generated by `Subsystem = %class` in my `logback.xml`):
```
! @7ahphlbml - Internal server error, for (GET) [/filewaf/robots.txt] ->

2019-01-16 19:45:02,100 | LogLevel = ERROR | ServiceID = FiLe | Version =  | Environment = DEV-Lieferant | HostName = LocalHost | Subsystem = play.api.http.DefaultHttpErrorHandler | Thread = play-dev-mode-akka.actor.default-dispatcher-40 | LogTrackingID =  | LogUserID =  | LogRequestID =  | Message =

! @7ahphlbml - Internal server error, for (GET) [/filewaf/robots.txt] ->

play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[NoSuchElementException: key not found: Id]]
        at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:323)
        at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:243)
        at play.api.http.DefaultHttpErrorHandler$.onServerError(HttpErrorHandler.scala:464)
        at play.core.server.Server$$anonfun$handleErrors$1$1.applyOrElse(Server.scala:96)
        at play.core.server.Server$$anonfun$handleErrors$1$1.applyOrElse(Server.scala:92)
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
        at play.core.server.Server$.getHandlerFor(Server.scala:115)
        at play.core.server.AkkaHttpServer.handleRequest(AkkaHttpServer.scala:267)
        at play.core.server.AkkaHttpServer.$anonfun$createServerBinding$1(AkkaHttpServer.scala:149)
        at akka.stream.impl.fusing.MapAsync$$anon$30.onPush(Ops.scala:1207)
Caused by: java.util.NoSuchElementException: key not found: Id
        at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:101)
        at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:99)
        at play.api.libs.typedmap.DefaultTypedMap.apply(TypedMap.scala:102)
        at play.api.mvc.RequestHeader.id(RequestHeader.scala:35)
        at play.api.mvc.RequestHeader.id$(RequestHeader.scala:35)
        at play.api.mvc.RequestHeaderImpl.id(RequestHeader.scala:352)
        at play.core.j.JavaHelpers.createJavaContext(JavaHelpers.scala:172)
        at play.core.j.JavaHelpers.createJavaContext$(JavaHelpers.scala:169)
        at play.core.j.JavaHelpers$.createJavaContext(JavaHelpers.scala:282)
        at play.core.j.JavaHelpers.withContext(JavaHelpers.scala:270)
```

**Comment by mkurz** [comment]
Can you post your whole `logback.xml`?

Also can you remove your `logback.xml` and try again?

**Comment by mkurz** [comment]
Alright, would it be possible that you upload your project to GitHub or provide it via a zip or gz file so I can debug myself?

**Comment by Squeng** [comment]
- cf. https://github.com/Squeng/Playverderber
- `PS C:\Users\Paul\Desktop\Playverderber> sbt`
- `[FiLe-App] $ h2-browser` -> note the empty DB (no PLAY_EVOLUTIONS table yet)
- `[FiLe-App] $ run -Dhttp.port=9000 -Dhttps.port=9443` and open https://localhost:9443/filewaf/bankers (accepting the Play-generated certificate)
- "Database 'default' needs evolution!" -> "Apply this script now!" => note the updated DB schema
- move `2.sql` from `public` to `conf\evolutions\default` and `ErrorHandler.java` to `app`
- re-open https://localhost:9443/filewaf/bankers
Et voilà:
```
[error]

! @7ai5a0k65 - Internal server error, for (GET) [/filewaf/bankers] ->

2019-01-17 14:33:48,081 | LogLevel = ERROR | ServiceID = FiLe | Version =  | Environment = DEV-Lieferant | HostName = LocalHost | Subsystem = play.api.http.DefaultHttpErrorHandler | Thread = play-dev-mode-akka.actor.default-dispatcher-56 | LogTrackingID =  | LogUserID =  | LogRequestID =  | Message =

! @7ai5a0k65 - Internal server error, for (GET) [/filewaf/bankers] ->

play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[NoSuchElementException: key not found: Id]]
        at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:323)
        at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:243)
        at play.api.http.DefaultHttpErrorHandler$.onServerError(HttpErrorHandler.scala:464)
        at play.core.server.Server$$anonfun$handleErrors$1$1.applyOrElse(Server.scala:96)
        at play.core.server.Server$$anonfun$handleErrors$1$1.applyOrElse(Server.scala:92)
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
        at play.core.server.Server$.getHandlerFor(Server.scala:115)
        at play.core.server.AkkaHttpServer.handleRequest(AkkaHttpServer.scala:267)
        at play.core.server.AkkaHttpServer.$anonfun$createServerBinding$1(AkkaHttpServer.scala:149)
        at akka.stream.impl.fusing.MapAsync$$anon$30.onPush(Ops.scala:1207)
Caused by: java.util.NoSuchElementException: key not found: Id
        at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:101)
        at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:99)
        at play.api.libs.typedmap.DefaultTypedMap.apply(TypedMap.scala:102)
        at play.api.mvc.RequestHeader.id(RequestHeader.scala:35)
        at play.api.mvc.RequestHeader.id$(RequestHeader.scala:35)
        at play.api.mvc.RequestHeaderImpl.id(RequestHeader.scala:352)
        at play.core.j.JavaHelpers.createJavaContext(JavaHelpers.scala:172)
        at play.core.j.JavaHelpers.createJavaContext$(JavaHelpers.scala:169)
        at play.core.j.JavaHelpers$.createJavaContext(JavaHelpers.scala:282)
        at play.core.j.JavaHelpers.withContext(JavaHelpers.scala:270)
```

**Comment by marcospereira** [comment]
I did some minimal debug so far, but since others may be handling this, let me share my findings:

I think the problem starts here:

https://github.com/playframework/playframework/blob/master/framework/src/play-server/src/main/scala/play/core/server/Server.scala#L101-L116

1. Play manages to get an application
2. It then calls the application request factory, which is [where the request id is populated](https://github.com/playframework/playframework/blob/1d232362264625658a3e2274dac6ddf19df0d037/framework/src/play/src/main/scala/play/api/mvc/request/RequestFactory.scala#L122-L128)
3. The [evolutions web command kicks in](https://github.com/playframework/playframework/blob/1d232362264625658a3e2274dac6ddf19df0d037/framework/src/play-jdbc-evolutions/src/main/scala/play/api/db/evolutions/ApplicationEvolutions.scala#L425-L444) and throws an exception since there is an evolution to apply
4. Play then needs to use the application error handler to properly handle the error. Since we have a Java error handler, we need to create an `Http.Context` which requires the request id. 
5. But we are not using the request from request factory anymore, so getting the request id fails and Play fallbacks to the default error handler.

The fix should happen in AkkaModelConversation, since we have a request id there already.

**Comment by mkurz** [comment]
@Squeng thanks for providing the project, I can reproduce the problem. Thanks!

@marcospereira Nice - will you provide the fix? I could test your pull request then ;)

**Comment by mkurz** [comment]
@Squeng we are working on a fixed here: #8965

**Comment by mkurz** [comment]
Fixed by #8965
